lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -47,125 +49,96 @@ const SessionAuth: React.FC<Props> = ({ children, ...props }) => {
+            if (!toSetContext.doesSessionExist && props.requireAuth === true) {
+                props.redirectToLogin();
+            } else {
+                setContext(toSetContext);
+            }
+        }
 
-        return setContext({
-            doesSessionExist: sessionExists,
-            jwtPayload,
-            userId,
-        });
+        setInitialContextAndMaybeRedirect();
+        return () => {
+            cancelUseEffect = true;
+        };
     }, []);
 
-    const onHandleEvent = useCallback(
-        async (event: RecipeEvent) => {
+    // subscribe to events on mount
+    useEffect(() => {
+        function onHandleEvent(event: RecipeEventWithSessionContext) {
             switch (event.action) {
                 case "SESSION_CREATED":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: true,
-                            userId: await getUserId(),
-                            jwtPayload: await getJWTPayloadSecurely(),
-                        });
-                    }
+                    setContext(event.sessionContext);
                     return;
                 case "REFRESH_SESSION":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: true,
-                            userId: context === undefined ? "" : context.userId,
-                            jwtPayload: await getJWTPayloadSecurely(),
-                        });
-                    }
+                    setContext(event.sessionContext);
                     return;
                 case "SIGN_OUT":
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: false,
-                            userId: "",
-                            jwtPayload: {},
-                        });
+                    if (props.requireAuth !== true) {
+                        setContext(event.sessionContext);
                     }
                     return;
                 case "UNAUTHORISED":
-                    // If there's onSessionExpired handler, use it without setting state...
-                    if (props.onSessionExpired !== undefined) {
-                        props.onSessionExpired();
-                        return;
-                    }
-
-                    // ...else fallback to default behaviour
-                    if (!hasParentProvider(parentSessionContext)) {
-                        setContext({
-                            doesSessionExist: false,
-                            userId: "",
-                            jwtPayload: {},
-                        });
+                    if (props.requireAuth === true) {
+                        if (props.onSessionExpired !== undefined) {
```

Comment:
It's possible to pass both `redirectToLogin` and `onSessionExpired`, while the former will be called only if `onSessionExpired` doesn't exist. This should be tested and documented.

Reply:
No.. this is expected behaviour.. And it's not surprising.. cause when you pass `onSessionExpired`, you expect that to be called in case of session expiry.. `redirectToLogin` is primarily meant to be used for when this component is mounted and no session exists and `requireAuth=true`.

We use `redirectToLogin` here as a fallback only. So this behaviour is intuitive and doesn't need to be documented..

Reply:
It might be expected for us, but I'm thinking about the users. If I were presented with component:
`<SessionAuth requireAuth={true} onSessionExpired={handler1} redirectToLogin={handler2} />`, I wouldn't think that `onSessionExpired` will prevent calling `redirectToLogin`. That's a matter of personal opinion, again, so hard to judge. For me it's complex, just look how many discussions we had on it and how much thinking went into designing this behaviour 😅

Reply:
So `onSessionExpired` will not be called if a session doesn't exist.. then `redirectToLogin` will be called. 

As the name suggests, `onSessionExpired` will only be called when a session has expired.. which is expected.

Isn't this also intuitive for the end user?